src/lua/mod.rs

```diff
@@ -218,61 +202,173 @@ fn thread_handle_message(sender: &Sender<LuaResponse>,
                 }
             }
+        }
+        return None;
+    }
+    return Some(table);
+}
+/// Converts a Json map into an AnyLuaValue
+pub fn json_to_lua(json: Json) -> AnyLuaValue {
+    match json {
+        Json::String(val)  => AnyLuaValue::LuaString(val),
+        Json::Boolean(val) => AnyLuaValue::LuaBoolean(val),
+        Json::F64(val)     => AnyLuaValue::LuaNumber(val),
+        Json::I64(val)     => AnyLuaValue::LuaNumber((val as i32) as f64),
+        Json::U64(val)     => AnyLuaValue::LuaNumber((val as u32) as f64),
+        Json::Null         => AnyLuaValue::LuaNil,
+        Json::Array(vals)  => {
+            let mut lua_arr = Vec::with_capacity(vals.len());
+            for (ix, val) in vals.into_iter().enumerate() {
+                lua_arr.push((AnyLuaValue::LuaNumber(ix as f64 + 1.0),
+                              json_to_lua(val)));
+            }
+            AnyLuaValue::LuaArray(lua_arr)
         },
+        Json::Object(vals) => {
+            let mut lua_table = Vec::with_capacity(vals.len());
+            for (key, val) in vals.into_iter() {
+                lua_table.push((AnyLuaValue::LuaString(key),
+                                json_to_lua(val)));
+            }
+            AnyLuaValue::LuaArray(lua_table)
+        }
+    }
+}
 
-        LuaQuery::SetValue { name: _name, val: _val } => {
-            panic!("thread: unimplemented LuaQuery::SetValue!");
-        },
+/// Converts an `AnyLuaValue` to a `Json`.
+///
+/// For an already-matched `LuaArray`, use `lua_array_to_json`.
+///
+/// For a `LuaArray` that should be mapped to a `JsonObject`,
+/// use `lua_object_to_json`.
+pub fn lua_to_json(lua: AnyLuaValue) -> Result<Json, ()> {
+    match lua {
+        AnyLuaValue::LuaNil => Ok(Json::Null),
+        AnyLuaValue::LuaString(val) => Ok(Json::String(val)),
+        AnyLuaValue::LuaNumber(val) => Ok(Json::F64(val)),
+        AnyLuaValue::LuaBoolean(val) => Ok(Json::Boolean(val)),
+        AnyLuaValue::LuaArray(arr) => lua_array_to_json(arr),
+        AnyLuaValue::LuaOther => Err(())
+    }
+}
 
-        LuaQuery::EmptyArray(_name) => {
-            panic!("thread: unimplemented LuaQuery::EmptyArray!");
-        },
+pub fn lua_array_to_json(arr: Vec<(AnyLuaValue, AnyLuaValue)>)
+                         -> Result<Json, ()> {
+    // Check if every key is a number
+    let mut counter = 0.0; // Account for first index?
 
-        _ => {
-            panic!("Unimplemented send type for lua thread!");
+    for &(ref key, ref _val) in &arr {
+        match *key {
+            AnyLuaValue::LuaNumber(num) => {
+                counter += num;
+            }
+            AnyLuaValue::LuaString(_) => {
+                break;
+            }
+            // Non-string keys are not allowed
+            _ => {
+                return Err(());
+            }
         }
     }
+
+    // Gauss' trick
+    let desired_sum = ((arr.len()) * (arr.len() + 1)) / 2;
+    if counter != desired_sum as f64 {
```

Comment:
If any of the indicies are missed, an Object is returned instead of an Array. Perhaps a better solution is to index those with a Null.